test/formal/checked_uint_mul_16.py

```diff
@@ -25,13 +25,17 @@
 	# cast to full n_bits values
 	X = BVUnsignedUpCast(X_short, n_bits)
 	Y = BVUnsignedUpCast(Y_short, n_bits)
+	product = MUL(X, Y)
 
 	# Constants
 	maxValue = BVUnsignedMax(type_bits, n_bits)
 
 	# Overflow check in YulUtilFunction::overflowCheckedIntMulFunction
-	overflow_check = AND(ISZERO(ISZERO(X)), GT(Y, DIV(maxValue, X)))
+	if type_bits == n_bits:
+		overflow_check = AND(ISZERO(ISZERO(Y)), ISZERO(EQ(X, DIV(product, Y))))
```

Comment:
This check has the same problem.

Reply:
I think there's two problems here: one is just that Z3 is surprisingly bad to handle this with larger bit sizes (probably due to the division).
For ``type_bits == 16`` and ``n_bits == 16``, it just reports "Unable to prove the rule", but does not provide a counterexample, which means, it just times out and cannot tell, if the rule is correct.

If you set the values even more conservatively to ``type_bits = 4`` and ``n_bits = 8`` above, it proves the rule for
``type_bits == 4 && n_bits == 8`` as well as for ``type_bits == 8 && n_bits == 8``.
However, it reports a counterexample for ``type_bits == 6 && n_bits == 8``. That's the tricky case, in which ``n_bits / 2 < type_bits < n_bits``... in our setting, that would for example be the multiplication of two ``uint160``s. For that case, we'll need to come up with a different check, since the current one actually doesn't work... if we don't find anything better, we could just keep the current checks for that, though.

Reply:
The signed case may even be more complicated :-).

Reply:
Interesting. Do you have the values of the counterexample for `type_bits == 6 && n_bits == 8`?

Reply:
For the record from chat: ``Model: [Y = 40, X = 7]``

 40 * 7 does overflow in 6-bit arithmetic, 40 * 7 = 280. But 280 modulo 256 (since we assume 8-bits in the ambient type) is 24, so product is 24. maxValue, however, in 6 bits is 63... so the check product > maxValue, i.e. 24 > 63, is false, even though you did overflow.